enhancements/security/bmc-ca-certificate-support.md

```diff
@@ -0,0 +1,294 @@
++      -----BEGIN CERTIFICATE-----
++      ......
++      ......
++      ......
++      -----END CERTIFICATE-----
++      "master-0.crt": |
++      -----BEGIN CERTIFICATE-----
++      ......
++      ......
++      ......
++      -----END CERTIFICATE-----
++      ......
+     ......
+     hosts:
+         - name: master-0
+           bmc:
+             disableCertificateVerification: false
+         ......
+     ......
+```
+
+Please note that these CA certificates will be merged into a single file within
+the Ironic container, rather than organized in the form defined by the API.
+Using a map for this field, as opposed to a string, improves maintainability.
+For instance, when updating CA certificates for certain nodes, users can
+distinguish them by the names they have set, reducing the risk of human errors.
+
+### Risks and Mitigations
+
+Users may incorrectly configure the CA certificates, causing the IPI
+installation to timeout. It may take some time for users to realize that the
+failure is due to incorrect certificate configuration.
+
+To mitigate this, we can perform a pre-validation of the SSL connection with
+the BMCs using the provided CA certificate before the installation begins.
+
+### Drawbacks
+
+- Introducing CA certificate support increases complexity to the installation
+  and management process.
+
+- Creating or deleting the CA certificate ConfigMap in an existing cluster
+  causes the Ironic container to restart, leading to temporary unavailability.
+
+- CA certificate pre-validation performed in installer is only for cluster
+  installation. There is no automatic pre-validation process for CA certificate
+  updates in an existing cluster.
+
+## Design Details
+
+### Test Plan
+
+TBD
+
+### Graduation Criteria
+
+This work will be GA immediately since there is no phasing possible or planned.
+
+#### Dev Preview -> Tech Preview
+
+N/A
+
+#### Tech Preview -> GA
+
+N/A
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+This enhancement should not affect the Upgrade/Downgrade Strategy.
+
+### Version Skew Strategy
+
+This enhancement requires the support of both `cluster-baremetal-operator`
+and `ironic-image`. The version of ironic-image is controlled by
+`cluster-baremetal-operator`, thus there will be no version skew.
+
+### Operational Aspects of API Extensions
+
+N/A
+
+#### Failure Modes
+
+1. Before the IPI installation begins, the `installer` will perform a
+   pre-validation using the provided CA certificate. If the validation fails,
+   an error will be reported immediately, indicating which baremetal BMC failed
+   the validation, and the installation will not proceed.
+
+2. If the pre-validation passes, but a certificate validation error occurs
+   during the installation process (e.g., due to the changes to the baremetal
+   BMC certificate or a man-in-the-middle attack), the user will not be able to
+   see the error message directly until timeout. However, if the user notices
```

Comment:
Is it so? I expect Ironic to throw an error when it cannot connect to the BMC. More specifically, the transition to `manageable` will not happen.

Reply:
Have you tested it with recent OpenShift? I know we do relay at least some errors to the user, maybe you're hitting a bug here?

Reply:
I have performed IPI deployment with openshift-baremetal-install in version `4.14.0-ec.4`. When there is a problem with the BMC certificate, an error will only occur in ironic, and this error will not be output to the terminal.

Error message in ironic container:
```
2023-08-03 05:26:49.518 1 ERROR ironic.drivers.modules.irmc.common [None req-fa76cc3f-27f6-46a9-854b-f6dd9dcd013e - - - - - -] Failed to get secure boot for node b1c4569f-caf3-4718-8bad-63425e70b228: scciclient.irmc.scci.SCCIClientError: HTTPSConnectionPool(host='192.168.30.27', port=443): Max retries exceeded with url: /rest/v1/Oem/eLCM/ProfileManagement/BiosConfig (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1129)'))
```
Debug information of the terminal:
```
DEBUG ironic_node_v1.openshift-master-host[1]: Still creating... [58m50s elapsed]     
DEBUG ironic_node_v1.openshift-master-host[1]: Still creating... [59m0s elapsed]      
DEBUG ironic_node_v1.openshift-master-host[1]: Still creating... [59m10s elapsed]     
DEBUG ironic_node_v1.openshift-master-host[1]: Still creating... [59m20s elapsed]     
DEBUG ironic_node_v1.openshift-master-host[1]: Still creating... [59m30s elapsed] 
```

Reply:
Okay, maybe we just automatically retry deploy failure.

Reply:
> I know we do relay at least some errors to the user

Sadly we do not.